Make Discover field chooser items keyboard accessible.

cjcenizal opened the pull request:
Partially addresses https://github.com/elastic/kibana/issues/11541

The general idea here is that you should be able to tab to interactive elements and use enter/space to interact with them (i.e. trigger a click handler). Ideally we'd use button elements, but because buttons can't contain other buttons, we have to make the containing div accessible by adding `tabindex="0"` and a keypress handler. See https://github.com/elastic/kibana/issues/11513 for more context.

snide commented:
👍  Looks good. Seems to work. Think to address that issue completely we'll need to deal with the caret expand bits in the content rows as well.

cjcenizal commented:
@snide Ah gotcha. BTW I just pushed a commit that also makes the links and +/- icons inside of the field chooser tabbable. I had to swap out the icons for buttons, and I used the MicroButton in the UI Framework, so there's a minor design change.

## Before

![image](https://cloud.githubusercontent.com/assets/1238659/25689725/a35b70d8-303f-11e7-8f00-087a79a09138.png)

## After

![image](https://cloud.githubusercontent.com/assets/1238659/25689728/a7b5d5b0-303f-11e7-9243-99865661ff06.png)

The contrast goes down quite a bit. We could revert this change and create custom buttons that preserve the original appearance. Thoughts?

Bargs commented on src/core_plugins/kibana/public/discover/components/field_chooser/discover_field.js:
```diff
@@ -95,6 +93,13 @@ app.directive('discoverField', function ($compile) {
+        if ([13, 32].includes(e.keyCode)) {
+          $scope.toggleDetails(field);
+        }
+      };
```
If we're going to need to do this often, do you think it would be worthwhile to create a custom attribute directive called `kbn-buttonPress` or something like that, which could handle this keycode checking and also do the preventDefault for the space bar on keyDown? It might make this a bit easier to implement in other components.

cjcenizal replied:
Thanks for bringing this up! I think it's a great idea but I'm a little stumped for the best place and name for this module. Suggestions?

Bargs replied:
`src/ui/public/directives`?

I suggested `kbn-buttonPress` because it is kind of similar to the angular event handling directives (`ng-click`, `ng-keyup` etc.) which we would be mimicking and a "buttonPress" (pressing space or enter on a button) is what we're trying to emulate.

cjcenizal replied:
Sounds good to me. How do you feel about `kbn-accessible-click`? I think it's good to emphasize that it's for aiding accessibility up-front (and I imagine we'll have other services and directives geared towards accessibility, so they'll be some cohesion here), and I think there's nice symmetry with `ng-click` since they both have the word "click" in them.

Bargs commented on src/core_plugins/kibana/public/discover/components/field_chooser/discover_field.html:
```diff
@@ -4,22 +4,24 @@
+      ng-bind="::field.display ? 'remove' : 'add'"
+      class="discover-sidebar-item-action kuiButton kuiButton--small kuiButton--primary"
+      data-test-subj="fieldToggle-{{::field.name}}"
+    ></button>
```
For some reason this button is not reacting to the space bar, even though it is a native button.

cjcenizal replied:
Which OS and browser are you using? When I tab to it and hit space, it's added to the selected fields list. I'm on Chrome on OS X.

cjcenizal replied:
Never mind, I can repro now that I implemented your preventDefault suggestion.

cjcenizal commented:
@Bargs Thanks for the great review. I believe I've addressed everything. Can you take another look?

Bargs commented on src/ui/public/accessibility/kbn_accessible_click.js:
```diff
@@ -0,0 +1,86 @@
+      }
+
+      element.on('keyup', e => {
+        // Support keyboard accessibility by toggling display on ENTER or SPACE keypress.
```
Comment seems out of date, should be something like 

```
// Support keyboard accessibility by emulating mouse clicks on ENTER or SPACE keypress.
```

Bargs commented on src/ui/public/accessibility/kbn_accessible_click.js:
```diff
@@ -0,0 +1,86 @@
+
+          if (!ngClick) {
+            throw new Error('kbnAccessibleClick requires a callback or ng-click defined on its element.');
+          }
```
Could you move this out of the event handler and into the main link function so that it fails early if ngClick is not specified?

Bargs commented on src/ui/public/accessibility/kbn_accessible_click.js:
```diff
@@ -0,0 +1,86 @@
+            throw new Error('kbnAccessibleClick requires a callback or ng-click defined on its element.');
+          }
+
+          ngClick(scope.$parent);
```
I hadn't considered using this directive in conjunction with ng-click, it's a great idea! You've got me wondering, do we even want to provide the ability to specify a separate event handler on this directive, or should we always require ng-click? My gut says the latter because it would ensure the interaction is always the same regardless of whether you're using a keyboard or mouse to interact with the element.

I also think we could avoid manually parsing the ng-click attribute if we simply fire a click event and allow ng-click to do its work as usual. This block could be simplified to:

```
      element.on('keyup', e => {
        if (accessibleClickKeys[e.keyCode]) {
          element.click();
        }
      });
```

One other question along this line of thought: you mentioned in https://github.com/elastic/kibana/issues/11513 that listening to the keyup event was necessary because "some browsers do not trigger onclick events for such elements when activated via the keyboard." Are there browsers that *do* trigger click events and if so will having an `ng-click` and an `kbn-accessible-click` on the same element cause the event handler to execute twice?

cjcenizal replied:
I was afraid of being too prescriptive by requiring an ng-click, but I guess you're right. That's sort of the whole idea, so I'll remove the ability to provide a handler.

Great suggestion on delegating to the ng-click via a click event, I'll implement your suggestion.

I did a little research but I couldn't find any information on which browsers may trigger click events on non-interactive elements. Hopefully we can recruit the help of an expert to help us with this (@jimgoodwin is already looking into this).

cjcenizal commented:
Thanks @Bargs, this is ready for your 👀  again.

Bargs commented on src/ui/public/accessibility/kbn_accessible_click.js:
```diff
@@ -0,0 +1,83 @@
+  return {
+    restrict: 'A',
+    scope: {
+      kbnAccessibleClick: '&',
```
I think this scope variable can be removed now.

Bargs commented on src/ui/public/accessibility/kbn_accessible_click.js:
```diff
@@ -0,0 +1,83 @@
+      }
+
+      // We're emulating a click action, so we should already have a regular click handler defined.
+      if (!$parse(attrs.ngClick)) {
```
It seems that `$parse(undefined)` returns a noop function, so this error never throws. `if (!attrs.ngClick)` should be sufficient.

Bargs commented on src/ui/public/accessibility/kbn_accessible_click.js:
```diff
@@ -0,0 +1,83 @@
+import { uiModules } from 'ui/modules';
+
+uiModules.get('kibana')
+.directive('kbnAccessibleClick', function ($parse) {
```
`$parse` will be unused once that usage below is removed.

cjcenizal commented:
@lukasolson Thanks for pointing out that behavior. Per our research and convo, links shouldn't activate on SPACE keypress, so we should be OK on that front.

@Bargs I've implemented your suggestions and added unit tests.

Could you both please take another look?

Bargs commented on src/ui/public/accessibility/__tests__/kbn_accessible_click.js:
```diff
@@ -0,0 +1,120 @@
+      const html = `<a ng-click="noop" kbn-accessible-click></a>`;
+      expect(() => {
+        $compile(html)($rootScope);
+      }).not.to.throwError(`kbnAccessibleClick doesn't need to be used on a link if it has a href attribute.`);
```
If you switch this to a regex it will blow up. It seems that it's not possible to test for **not** throwing a specific exception, so you should just pass 0 args here.

Bargs commented on src/ui/public/accessibility/__tests__/kbn_accessible_click.js:
```diff
@@ -0,0 +1,120 @@
+    const child = angular.element(`<button></button>`);
+    element.append(child);
+    const e = angular.element.Event('keydown'); // eslint-disable-line new-cap
+    e.which = ENTER_KEY;
```
Why are you setting `e.which` here when you were setting `e.keyCode` in the tests above?

Bargs commented on src/ui/public/accessibility/__tests__/kbn_accessible_click.js:
```diff
@@ -0,0 +1,120 @@
+    });
+  });
+
+  it(`doesn't call ng-click when the originating element is a child`, () => {
```
This doesn't seem to be working correctly. This check happens in the `keydown` [event handler](https://github.com/elastic/kibana/blob/059b6975b5874174379648a9e6cd3d6c719a190d/src/ui/public/accessibility/kbn_accessible_click.js#L34-L34). But the click gets fired in the `keyup` [event handler](https://github.com/elastic/kibana/blob/059b6975b5874174379648a9e6cd3d6c719a190d/src/ui/public/accessibility/kbn_accessible_click.js#L75-L75). I set a breakpoint in the `keyup` event handler and it is still getting fired even if the event originated on a child. I think the only reason we're not noticing this with the "Add" button is that the element immediately gets destroyed. 

If you switch line 115 from `keydown` to `keyup` and line 116 from `e.which` to `e.keyCode` you'll see the test fail.

Bargs commented on src/ui/public/accessibility/__tests__/kbn_accessible_click.js:
```diff
@@ -0,0 +1,120 @@
+    child.trigger(e);
+    expect(scope.handleClick.callCount).to.be(0);
+  });
+});
```
Could you also add one more test for the preventDefault on `keyDown` for spaces?

cjcenizal commented on src/ui/public/accessibility/__tests__/kbn_accessible_click.js:
```diff
@@ -0,0 +1,120 @@
+      const html = `<button kbn-accessible-click></button>`;
+      expect(() => {
+        $compile(html)($rootScope);
+      }).to.throwError(`kbnAccessibleClick doesn't need to be used on a button.`);
```
Damn, we have tests elsewhere that pass strings into `throwError`. Thanks for letting me know.

cjcenizal commented:
@Bargs OK man chickity check it. Thanks for the critique.

Bargs commented on src/ui/public/accessibility/__tests__/kbn_accessible_click.js:
```diff
@@ -0,0 +1,135 @@
+      const child = angular.element(`<button></button>`);
+      element.append(child);
+      const e = angular.element.Event('keyup'); // eslint-disable-line new-cap
+      e.keyCode = ENTER_KEY;
```
I think you mean `SPACE_KEY` here?

cjcenizal replied:
Great catch, as always.

cjcenizal commented:
jenkins, test this

cjcenizal commented:
jenkins, test this
